### PR TITLE
Added nested categories, capybara tests and fixed bugs.

### PR DESCRIPTION
**Pull request by Zawgor24**
List of features

**Tests:**
  Fixed specs of models.
  Added:
     Support folder and some helper methods.
     Features folder to capybara specs.

**Categories:**
  Added: 
     categories to posts.
     nested categories via Ancestry gem.
     Front-end to pages with categories.
etc.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -25,6 +25,7 @@ gem 'popper_js', '~> 1.12.9'
 gem 'puma', '~> 3.7'
 gem 'rails', '~> 5.1.4'
 gem 'rmagick'
+gem 'rspec'
```
Comment (IlyaUmanets) [inline Gemfile]: rspec-rails

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -1,7 +1,20 @@
...
   protect_from_forgery with: :exception
 
   before_action :authenticate_user!
+
+  def configure_permitted_parameters
+    devise_parameter_sanitizer.permit(:sign_up) do |u|
+      u.permit(:email, :password,
+        :password_confirmation, :remember_me, :first_name, :last_name)
+    end
+    devise_parameter_sanitizer.permit(:account_update) do |u|
```
Comment (IlyaUmanets) [inline app/controllers/application_controller.rb]: put empy line after `end`

File: app/controllers/comments_controller.rb [hunk 3]
```diff
@@ -11,9 +11,9 @@ def new
   def create
     @comment = @commentable.comments.new(comment_params_with_user)
     if @comment.save
-      redirect_to @commentable
+      redirect_to request.referer, notice: 'Comment successfully created'
```
Comment (IlyaUmanets) [inline app/controllers/comments_controller.rb]: move this notice to `I18n`

File: app/controllers/comments_controller.rb [hunk 4]
```diff
@@ -11,9 +11,9 @@ def new
...
     else
-      render :new
+      redirect_to request.referer, notice: 'Enter valid data'
```
Comment (IlyaUmanets) [inline app/controllers/comments_controller.rb]: same

File: app/controllers/comments_controller.rb [hunk 5]
```diff
@@ -30,7 +30,7 @@ def update
   def destroy
     @comment.destroy
 
-    redirect_to @comment.commentable
+    redirect_to request.referer, notice: 'Comment deleted'
```
Comment (IlyaUmanets) [inline app/controllers/comments_controller.rb]: same

File: app/controllers/posts_controller.rb [hunk 6]
```diff
@@ -1,10 +1,13 @@
 # frozen_string_literal: true
 
 class PostsController < ApplicationController
+  before_action :find_category, only: %i[index]
   before_action :find_post, only: %i[show edit update destroy]
 
   def index
-    @posts = Post.where(category_id: find_category)
+    @posts = Post.where(category_id: @category.subtree_ids).order(
+      'updated_at DESC'
```
Comment (IlyaUmanets) [inline app/controllers/posts_controller.rb]: `.order(updated_at: :desc)`

File: app/controllers/posts_controller.rb [hunk 7]
```diff
@@ -17,9 +20,10 @@ def create
     @post = current_user.posts.new(post_params)
 
     if @post.save
-      redirect_to @post
+
+      redirect_to @post, notice: 'Post successfully created'
```
Comment (IlyaUmanets) [inline app/controllers/posts_controller.rb]: same

File: app/controllers/posts_controller.rb [hunk 8]
```diff
@@ -17,9 +20,10 @@ def create
...
     else
-      render :new
+      render :new, notice: 'Enter valid data'
```
Comment (IlyaUmanets) [inline app/controllers/posts_controller.rb]: same

File: app/controllers/posts_controller.rb [hunk 9]
```diff
@@ -36,7 +40,7 @@ def update
   def destroy
     @post.destroy
 
-    redirect_to root_path
+    redirect_to root_path, notice: 'Post deleted'
```
Comment (IlyaUmanets) [inline app/controllers/posts_controller.rb]: same

File: app/controllers/posts_controller.rb [hunk 10]
```diff
@@ -46,7 +50,7 @@ def find_post
   end
 
   def find_category
-    Category.find(params[:category_id])
+    @category = Category.find(params[:category_id])
```
Comment (IlyaUmanets) [inline app/controllers/posts_controller.rb]: `@category || =`

File: app/helpers/application_helper.rb [hunk 11]
```diff
@@ -4,4 +4,27 @@ module ApplicationHelper
...
+    end
+
+    result << content_tag(:li, category.name, class: 'breadcrumb-item active')
+
+    safe_join(result)
+  end
+
+  def flash_class(level)
+    case level
+    when :notice then 'alert alert-info'
```
Comment (IlyaUmanets) [inline app/helpers/application_helper.rb]: `[notice: 'alert alert-info', success: 'alert alert-success', ....][level]`

File: app/models/category.rb [hunk 12]
```diff
@@ -1,6 +1,8 @@
 # frozen_string_literal: true
 
 class Category < ApplicationRecord
+  has_ancestry
```
Comment (IlyaUmanets) [inline app/models/category.rb]: Put any `gem` stuff after `Rails` validations, associations etc.

File: app/models/user.rb [hunk 13]
```diff
@@ -13,13 +13,14 @@ class User < ApplicationRecord
   has_many :articles, dependent: :destroy
   has_many :posts, dependent: :destroy
 
+  mount_uploader :avatar, ImageUploader
```
Comment (IlyaUmanets) [inline app/models/user.rb]: same

File: spec/features/category_spec.rb [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+
+RSpec.describe 'Category', type: :feature do
+  let(:user) { create(:user) }
+
+  subject { page }
+
+  before { sign_in(user) }
+
+  describe '#show' do
+    context "when user isn't authorized" do
```
Comment (IlyaUmanets) [inline spec/features/category_spec.rb:13]: `''...."` -> `'...'`. Use single quotes.

File: spec/features/comment_spec.rb [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+
+  subject { page }
+
+  before { sign_in(user) }
+
+  describe '#create at post' do
+    context "when user isn't authorized" do
+      let(:user) { build(:user) }
+
+      scenario { expect(page.has_no_content?(I18n.t(:log_out))).to eq(true) }
```
Comment (IlyaUmanets) [inline spec/features/comment_spec.rb]: `.to be_truthy`

File: spec/features/post_spec.rb [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+
+  subject { page }
+
+  before { sign_in(user) }
+
+  describe '#create' do
+    context 'when user unauthorized' do
+      let(:user) { build(:user) }
+
+      scenario { expect(page.has_no_content?(I18n.t(:title))).to eq(true) }
```
Comment (IlyaUmanets) [inline spec/features/post_spec.rb]: same

**Review by IlyaUmanets — CHANGES_REQUESTED** [review]

File: Gemfile (hunk 1 above)
Comment (Zawgor24) [reply Gemfile]: Ok.

File: app/controllers/application_controller.rb (hunk 2 above)
Comment (Zawgor24) [reply app/controllers/application_controller.rb]: Done.

File: app/controllers/comments_controller.rb (hunk 3 above)
Comment (Zawgor24) [reply app/controllers/comments_controller.rb]: Understand, fixed.

File: app/controllers/posts_controller.rb (hunk 6 above)
Comment (Zawgor24) [reply app/controllers/posts_controller.rb]: OK, done.

File: app/controllers/posts_controller.rb (hunk 7 above)
Comment (Zawgor24) [reply app/controllers/posts_controller.rb]: Done.

File: app/controllers/posts_controller.rb (hunk 10 above)
Comment (Zawgor24) [reply app/controllers/posts_controller.rb]: Fixed.

File: app/helpers/application_helper.rb (hunk 11 above)
Comment (Zawgor24) [reply app/helpers/application_helper.rb]: Added notices into hash.

File: app/models/category.rb (hunk 12 above)
Comment (Zawgor24) [reply app/models/category.rb]: Got it. Corrected.

File: app/models/user.rb (hunk 13 above)
Comment (Zawgor24) [reply app/models/user.rb]: Done.

File: app/controllers/users_controller.rb [hunk 17]
```diff
@@ -5,6 +5,20 @@ class UsersController < ApplicationController
 
   def show; end
 
+  def new
+    @user = User.new
+  end
+
+  def create
+    @user = User.create(user_params)
```
Comment (IlyaUmanets) [inline app/controllers/users_controller.rb]: `User.new`

File: app/views/devise/registrations/new.html.slim [hunk 18]
```diff
@@ -29,8 +43,13 @@
...
-          .form-group style=("border-top: 1px solid #999; padding-top:20px") 
+          .form-group 
             .col-md-offset-3.col-md-9
               button#btn-fbsignup.btn.btn-primary type="button" 
                 i.icon-facebook
                 |   Sign Up with Facebook
+          .form-group style=("border-top: 1px solid #999; padding-top:20px") 
+            .col-md-12.control
+              | Click 
+              = link_to "Log in", new_session_path(resource_name)
```
Comment (IlyaUmanets) [inline app/views/devise/registrations/new.html.slim]: single quotes. Check all html files please.

File: spec/features/post_spec.rb [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+      let(:user) { build(:user) }
+
+      scenario { expect(page.has_no_content?(I18n.t(:title))).to be_truthy }
+    end
+
+    context 'when user authorized' do
+      before { visit new_user_post_path(user) }
+
+      context 'when valid data' do
+        scenario 'will create a post' do
```
Comment (IlyaUmanets) [inline spec/features/post_spec.rb]: `creates a post`

File: spec/features/post_spec.rb [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+            'support', 'fixtures', 'default-post-image.jpg'))
+
+          click_button 'Submit'
+
+          is_expected.to have_content(post[:title])
+        end
+      end
+
+      context 'when invalid data' do
+        scenario "won't create a post" do
```
Comment (IlyaUmanets) [inline spec/features/post_spec.rb]: `does not create a post`

File: spec/features/profile_spec.rb [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+      authorize
+      click_link 'Edit profile'
+    end
+
+    context 'when valid data' do
+      let(:user_attrs) { attributes_for(:user, :full) }
+
+      scenario { is_expected.to have_content(I18n.t('profiles.edit.editing')) }
+
+      scenario 'will edit user information' do
```
Comment (IlyaUmanets) [inline spec/features/profile_spec.rb]: `edits user information`

**Review by IlyaUmanets — CHANGES_REQUESTED** [review]

**Review by IlyaUmanets — APPROVED** [review]

**Pull request merged**